src/custom_nodes/common/buffersqueue.cpp

```diff
@@ -0,0 +1,52 @@
+//*****************************************************************************
+// Copyright 2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "buffersqueue.hpp"
+#include <iostream>
+
+namespace ovms {
+namespace custom_nodes_common {
+BuffersQueue::BuffersQueue(size_t singleBufferSize, int streamsLength) :
+    Queue(streamsLength),
+    singleBufferSize(singleBufferSize),
+    size(singleBufferSize * streamsLength),
+    memoryPool(new char[size]) {
+    if (nullptr == memoryPool) {
+        throw std::bad_alloc();
+    }
+    for (int i = 0; i < streamsLength; ++i) {
+        inferRequests.push_back(memoryPool + i * singleBufferSize);
+    }
+}
+BuffersQueue::~BuffersQueue() {
+    delete[] memoryPool;
+}
+void* BuffersQueue::getBuffer() {
+    // can be easily switched to async version if need arise
+    return getInferRequest(getIdleStream().get());
```

Comment:
Usage of this buffer queue in custom nodes will cause deadlocks in OVMS since custom node execution is single threaded and multiple custom nodes can be used in the same pipeline.

Reply:
Yes if you get enough buffers blocked it will block execution, but you can still allocate from heap but you pay with perf. That's why release implementation will be more complicated.

Reply:
After discussion agreed to leave up to the user of BuffersQueue API to either get buffer straight away or allocate on their own